### PR TITLE
AlgoCash 1st Milestone Delivery

### PR DESCRIPTION
**Pull request by Dohkooo**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/217 < please fill this in with the PR number of your application.

### CODE REVIEW

**Comment by Noc2** [comment]
Thanks for the delivery. We will look into it as soon as possible.

**Comment by Noc2** [comment]
@Dohkooo Thanks again for the delivery. I started to look into it and have a few questions, [see](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/AlgoCash_1_Noc2.md). In general it’s really hard to follow your code, since there is no inline documentation or a description of exactly how the mechanism works (besides the text which you copied from the grant application). My biggest questions so far are:
- The demo video shows an ALC and ALB contract, which are also part of the contract, but they aren’t part of the repo as far as I can see. Where can I find these? 
- Why did you implement these tests: https://github.com/ReserveLabs/AlgoCash/blob/6ef24ec93eb19aaf5b7ceb74635bc0d6a3da1e8d/contracts/treasury/lib.rs#L270
- Why can I choose the [target price](https://github.com/ReserveLabs/AlgoCash/blob/6ef24ec93eb19aaf5b7ceb74635bc0d6a3da1e8d/contracts/treasury/lib.rs#L149) when I buy bonds? Is there an auction?

**Comment by Dohkooo** [comment]
@Noc2 Thank you for taking time for this review.

>In general it’s really hard to follow your code, since there is no inline documentation or a description of exactly how the mechanism works (besides the text which you copied from the grant application). 

We will add inline comments to make code more readable and we will modify the README file to better illustrate the components and working mechanism. Once done, I will revert and comment here.

>The demo video shows an ALC and ALB contract, which are also part of the contract, but they aren’t part of the repo as far as I can see. Where can I find these?

Both of them are instantiated from [asset contract](https://github.com/ReserveLabs/AlgoCash/blob/main/contracts/asset/lib.rs) .

> Why did you implement these tests: https://github.com/ReserveLabs/AlgoCash/blob/6ef24ec93eb19aaf5b7ceb74635bc0d6a3da1e8d/contracts/treasury/lib.rs#L270

They are useless and we've deleted them.

>Why can I choose the target price when I buy bonds? Is there an auction?

You can set the highest price you want to buy bonds. If the price is higher than <targer price>, your purchase will fail.

>It would be nice if they could also have shown how to deploy the contracts.

We have a [deployment script](https://github.com/ReserveLabs/AlgoCash/blob/main/scripts/deploy.ts) to deploy the contracts and you could run npx redspot run scripts/deploy.ts to deploy them.

>This seems to be based on basis or more specifically basiscash-protocol. What’s the difference between this and the existing implementation for example on ethereum, which clearly doesn’t work, see https://www.coingecko.com/en/coins/basis-cash. (decreased demand)

Yes, we took basis' mechanism for our reference in our initial design. And our 1st milestone delivery is to implement the contractionary and expansionary policies. 

I believe the reasons why basis mechanism doesn't work are 1. when Shares price goes down, most users will stake them into boardroom and redeem cash to eliminate the loss, which cause the cash expansion and cash price goes down too fast. 2. when cash price goes below the threshold, few users believe it will revert and buy bonds. 

So in the 2nd milestone, we will add some DeFi protocol into AlgoCash ecosystem. It will offer users interst and profit  by locking certain amount of shares and bonds according to their issurance. Thus, making less shares staking in the boardroom and more users are interested in bonds.

**Comment by Dohkooo** [comment]
We have edited the README file and added more inline documentation.

**Comment by Noc2** [comment]
Thanks for the update and detailed reply. I see you are mostly just using the basis cash implementation. In this case the following might [be problematic](https://github.com/ReserveLabs/AlgoCash/blob/main/contracts/asset/lib.rs#L337). Why do you want to have the burn functionality for the cash token? Do you know if basis cash was audited? In any case, I’m happy to tell you the milestone is a pass and will forward your invoice internally.

**Comment by Dohkooo** [comment]
Thanks again for your time.

>Why do you want to have the burn functionality for the cash token? 

After you buy bonds, your cahes are burnt to make the contractionary.

I've seen the audit report from the website link of basis token. However it returns a 404 error now.

**Comment by Dohkooo** [comment]
Hi David,


May I know any update about the invoice, please?


Thank you,
Dohkooo




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "w3f/Grant-Milestone-Delivery"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年8月6日(星期五) 晚上8:17
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [w3f/Grant-Milestone-Delivery] AlgoCash 1st Milestone Delivery (#246)





 
Thanks for the update and detailed reply. I see you are mostly just using the basis cash implementation. In this case the following might be problematic. Why do you want to have the burn functionality for the cash token? Do you know if basis cash was audited? In any case, I’m happy to tell you the milestone is a pass and will forward your invoice internally.
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Noc2** [comment]
Hi, thanks for chasing us. I just sent the invoice team a reminder.

**Comment by RouvenP** [comment]
Hi @Dohkooo apologies for the delay. We're facing a technical issue at the moment but should be able to send the payment tomorrow - I will keep you posted.
